### ISSUE TITLE
Docs for `new_without_default` don't mention `pub` requirement

### ISSUE THREAD

**Issue by nsunderland1** [rust-lang/rust-clippy#8415]
### Description

The [docs](https://rust-lang.github.io/rust-clippy/master/#new_without_default) for the `new_without_default` lint include the following example:

```rust
struct Foo(Bar);

impl Foo {
    fn new() -> Self {
        Foo(Bar::new())
    }
}
```

If we run Clippy against this example (plus a definition for `Bar`), we get no warnings for `new_without_default`, see the [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b9418408c963904581f81c1043c3739d).

This seems to be because the warning will only trigger if both the type and the `new` method are `pub`. See [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d403757c51e896afb1e1a731767d39ef) similar example which does get the `new_without_default` warning.

The docs should be clear on this `pub` requirement, and the example should reflect this.

### Version

```text
rustc 1.58.1 (db9d1b20b 2022-01-20)
binary: rustc
commit-hash: db9d1b20bba1968c1ec1fc49616d4742c1725b4b
commit-date: 2022-01-20
host: x86_64-unknown-linux-gnu
release: 1.58.1
LLVM version: 13.0.0
```

### Additional Labels

@rustbot label +A-documentation
